src/Compressors.jl

```diff
@@ -0,0 +1,414 @@
+    end
+
+    return nothing
+end
+
+"""
+    right_mat_mul_dimcheck(C::AbstractMatrix, A::AbstractMatrix), S::CompressorRecipe)
+
+$(comp_method_description[:mul_check] * " from the right.") 
+
+### Arguments
+- $(comp_arg_list[:C]) 
+- $(comp_arg_list[:A]) 
+- $(comp_arg_list[:compressor_recipe])
+
+# Outputs
+- `Nothing`. 
+"""
+function right_mat_mul_dimcheck(C::AbstractMatrix, A::AbstractMatrix, S::CompressorRecipe)
+    s_rows, s_cols = size(S)
+    a_rows, a_cols = size(A)
+    c_rows, c_cols = size(C)
+    if a_cols != s_rows
+        throw(DimensionMismatch("Matrix A has $a_cols columns while S has $s_rows rows."))
+    elseif c_cols != s_cols
+        throw(DimensionMismatch("Matrix C has $c_cols columns while S has $s_cols columns.")
+             )
+    elseif c_rows != a_rows
+        throw(DimensionMismatch("Matrix C has $c_rows rows while A has $a_rows rows."))
+    end
+
+    return nothing
+end
+
+"""
+    vec_mul_dimcheck(z::AbstractMatrix, S::CompressorRecipe, y::AbstractMatrix)
+
+$(comp_method_description[:mul_check] * " with a vector.") 
+
+### Arguments
+- $(comp_arg_list[:z]) 
+- $(comp_arg_list[:compressor_recipe])
+- $(comp_arg_list[:y]) 
+
+# Outputs
+- `Nothing`. 
+"""
+function vec_mul_dimcheck(z::AbstractVector, S::CompressorRecipe, y::AbstractVector)
+    s_rows, s_cols = size(S)
+    len_y = size(y, 1)
+    len_z = size(z, 1)
+    if len_y != s_cols
+        throw(
+            DimensionMismatch(
+                "Vector y is of dimension $len_y while S has $s_cols columns."
+            )
+        )
+    elseif len_z != s_rows
+        throw(
+            DimensionMismatch("Vector z is of dimension $len_z while S has $s_rows rows.")
+        )
+    end
+
+    return nothing
+end
+# Implement the * operator  for matrix matrix multiplication
+function mul!(C::AbstractArray, S::CompressorRecipe, A::AbstractArray, alpha::Float64, 
+    beta::Float64)
+    throw(ArgumentError("No method `mul!` defined for ($(typeof(C)), $(typeof(S)), \
+    $(typeof(A)), $(typeof(alpha)), $(typeof(beta)))."))
+    return nothing
+end
+
+function mul!(C::AbstractArray, A::AbstractArray, S::CompressorRecipe, alpha::Float64, 
+    beta::Float64)
+    throw(ArgumentError("No method `mul!` defined for ($(typeof(C)), $(typeof(A)), \
+    $(typeof(S)), $(typeof(alpha)), $(typeof(beta)))."))
+    return nothing 
+end
+
+function (*)(S::CompressorRecipe, v::AbstractVector)
+    s_rows = size(S, 1)
+    output = zeros(s_rows)
+    vec_mul_dimcheck(output, S, v)
+    mul!(output, S, v, 1.0, 0.0)
+    return output
+end
+
+function mul!(x::AbstractVector, S::CompressorRecipe, y::AbstractVector)
+    mul!(x, S, y, 1.0, 0.0)
+    return nothing
+end
+
+# Implement the * operator for matrix matrix multiplication
+# The left multiplication version
+function (*)(S::CompressorRecipe, A::AbstractMatrix)
+    s_rows = size(S, 1)
+    a_cols = size(A, 2)
+    C = zeros(s_rows, a_cols)
```

Comment:
@nathanielpritchard don't you need to specify the type of C? Or are you making it a dense float array by default. At least maybe C should inherit the number type (float, double, integer, etc)